superset/migrations/env.py

```diff
@@ -42,8 +42,8 @@
         "SQLite Database support for metadata databases will \
         be removed in a future version of Superset."
     )
-decoded_uri = urllib.parse.unquote(DATABASE_URI)
-config.set_main_option("sqlalchemy.url", decoded_uri)
+escaped_uri = DATABASE_URI.replace('%', '%%')
```

Comment:
This looks like it might break existing functionality. Should we instead do the replace on the "decoded_url" var?

Reply:
@craig-rueda You mean use the same var name? Or apply the replace on top of `urllib.parse.unquote(DATABASE_URI)`.

Reply:
Like this:

```
decoded_uri = urllib.parse.unquote(DATABASE_URI)
decoded_uri = decoded_uri.replace('%', '%%')
config.set_main_option("sqlalchemy.url", decoded_uri)
```

So basically just add the replace in there. 

Note that if your PWD has a `%` symbol in it, you may need to "enquote" it going in, i.e. `some_pass_with_%_in_it` would need to be url encoded as `some_pass_with_%25_in_it`

Reply:
Ahhh.. I got it now. Sorry I should have tried to explain in detail the need for this change. 

Example : Password with @ in them super2p@ss. As suggested in [SQLAlechemy Docs](https://docs.sqlalchemy.org/en/14/core/engines.html#escaping-special-characters-such-as-signs-in-passwords)  if the password is either escaped and passed to superset or in superset_config.py if the user escapes the special char using ```urllib.parse.quote_plus("super2p@ss")``` alembic's env.py is going to unquote it and we end up with "super2p@ss" which will not be parsed by psycopg2.

1. user passes  -> ```"super2p@ss"```
2. superset_config.py parsers -> ```"super2p%40ss"```
3. superset_config.py sets DATABASE_URI -> ```"postgresql+psycopg2://user:super2p%40ss@localhost:5432/superset"```
4. env.py decodes DATABASE_URI into decoded_uri -> ```"postgresql+psycopg2://user:super2p@ss@localhost:5432/superset"```

And superset app starts up, but the migration fails with the error "psycopg2.OperationalError: could not translate host name "ss@localhost" to address" as it splits at symbol "@".

Just by removing the unquote in env.py will not fix the problem as the password is being set as a config, which uses configparser hence the "%" has to be double escaped into %% for it to be valid.

so the flow becomes..
1. user passes  -> ```"super2p@ss"```
2. superset_config.py parsers -> ```"super2p%40ss"```
3. superset_config.py sets DATABASE_URI -> ```"postgresql+psycopg2://user:super2p%40ss@localhost:5432/superset"```
4. env.py doubles escapes DATABASE_URI into escaped_uri -> ```"postgresql+psycopg2://user:super2p%%40ss@localhost:5432/superset"```

References:
https://github.com/psycopg/psycopg2/discussions/1546
https://github.com/sqlalchemy/alembic/issues/700#issuecomment-636036888

Reply:
Ok, makes sense. From what it looks like here, we're just trying to address an errant `@` from finding its way into the connection string.

Hoping this doesn't break anyone 🙏 

CC - @john-bodley @dpgaspar  ??